lib/charms/layer/kubernetes_common.py

```diff
@@ -933,3 +934,199 @@ def get_secret_password(username):
             if username == secret_user:
                 return password
     return None
+
+
+def get_node_ip():
+    """Determines the preferred NodeIP value for this node."""
+    cidr = cluster_cidr()
+    if not cidr:
+        return None
+    if is_ipv6_preferred(cidr):
+        return get_ingress_address6("kube-control")
+    else:
+        return get_ingress_address("kube-control")
+
+
+def merge_kubelet_extra_config(config, extra_config):
+    """Updates config to include the contents of extra_config. This is done
+    recursively to allow deeply nested dictionaries to be merged.
+
+    This is destructive: it modifies the config dict that is passed in.
+    """
+    for k, extra_config_value in extra_config.items():
+        if isinstance(extra_config_value, dict):
+            config_value = config.setdefault(k, {})
+            merge_kubelet_extra_config(config_value, extra_config_value)
+        else:
+            config[k] = extra_config_value
+
+
+def workaround_lxd_kernel_params():
+    """
+    Workaround for kubelet not starting in LXD when kernel params are not set
+    to the desired values.
+    """
+    if host.is_container():
+        hookenv.log("LXD detected, faking kernel params via bind mounts")
+        root_dir = "/root/cdk/lxd-kernel-params"
+        os.makedirs(root_dir, exist_ok=True)
+        # Kernel params taken from:
+        # https://github.com/kubernetes/kubernetes/blob/v1.22.0/pkg/kubelet/cm/container_manager_linux.go#L421-L426
+        # https://github.com/kubernetes/kubernetes/blob/v1.22.0/pkg/util/sysctl/sysctl.go#L30-L64
+        params = {
+            "vm.overcommit_memory": 1,
+            "vm.panic_on_oom": 0,
+            "kernel.panic": 10,
+            "kernel.panic_on_oops": 1,
+            "kernel.keys.root_maxkeys": 1000000,
+            "kernel.keys.root_maxbytes": 1000000 * 25,
+        }
+        for param, param_value in params.items():
+            fake_param_path = root_dir + "/" + param
+            with open(fake_param_path, "w") as f:
+                f.write(str(param_value))
+            real_param_path = "/proc/sys/" + param.replace(".", "/")
+            host.fstab_add(fake_param_path, real_param_path, "none", "bind")
+        subprocess.check_call(["mount", "-a"])
+    else:
+        hookenv.log("LXD not detected, not faking kernel params")
```

Comment:
Removed [here](https://github.com/charmed-kubernetes/charm-kubernetes-worker/pull/103/files#diff-1837fd4ad476d5c87c628d68b3c461fc5c03d6aeafc928e910c69aa6ae267f85L1517-L1545) and added to this PR